pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go

```diff
@@ -21,68 +20,74 @@ type AzureTokenProvider interface {
-	authParams *plugins.JwtTokenAuth
+	cfg         *setting.Cfg
+	credentials azcredentials.AzureCredentials
+	scopes      []string
 }
 
-func NewAzureAccessTokenProvider(cfg *setting.Cfg, authParams *plugins.JwtTokenAuth) *tokenProviderImpl {
+func NewAzureAccessTokenProvider(cfg *setting.Cfg, credentials azcredentials.AzureCredentials, scopes []string) *tokenProviderImpl {
 	return &tokenProviderImpl{
-		cfg:        cfg,
-		authParams: authParams,
+		cfg:         cfg,
+		credentials: credentials,
+		scopes:      scopes,
 	}
 }
 
 func (provider *tokenProviderImpl) GetAccessToken(ctx context.Context) (string, error) {
 	var credential TokenCredential
 
-	if provider.isManagedIdentityCredential() {
+	// TODO: Move to provider initialization and reuse for each GetAccessToken call
+	switch c := provider.credentials.(type) {
+	case *azcredentials.AzureManagedIdentityCredentials:
 		if !provider.cfg.Azure.ManagedIdentityEnabled {
 			err := fmt.Errorf("managed identity authentication is not enabled in Grafana config")
 			return "", err
 		} else {
-			credential = provider.getManagedIdentityCredential()
+			credential = provider.getManagedIdentityCredential(c)
 		}
-	} else {
-		credential = provider.getClientSecretCredential()
+	case *azcredentials.AzureClientSecretCredentials:
+		credential = provider.getClientSecretCredential(c)
+	default:
+		err := fmt.Errorf("credentials of type '%s' not supported by authentication provider", c.AzureAuthType())
+		return "", err
 	}
 
-	accessToken, err := azureTokenCache.GetAccessToken(ctx, credential, provider.authParams.Scopes)
+	accessToken, err := azureTokenCache.GetAccessToken(ctx, credential, provider.scopes)
 	if err != nil {
 		return "", err
 	}
 
 	return accessToken, nil
 }
 
-func (provider *tokenProviderImpl) isManagedIdentityCredential() bool {
-	authType := strings.ToLower(provider.authParams.Params["azure_auth_type"])
-	clientId := provider.authParams.Params["client_id"]
-
-	// Type of authentication being determined by the following logic:
-	// * If authType is set to 'msi' then user explicitly selected the managed identity authentication
-	// * If authType isn't set but other fields are configured then it's a datasource which was configured
-	//   before managed identities where introduced, therefore use client secret authentication
-	// * If authType and other fields aren't set then it means the datasource never been configured
-	//   and managed identity is the default authentication choice as long as managed identities are enabled
-	return authType == "msi" || (authType == "" && clientId == "" && provider.cfg.Azure.ManagedIdentityEnabled)
-}
-
-func (provider *tokenProviderImpl) getManagedIdentityCredential() TokenCredential {
-	clientId := provider.cfg.Azure.ManagedIdentityClientId
-
-	return &managedIdentityCredential{clientId: clientId}
+func (provider *tokenProviderImpl) getManagedIdentityCredential(credentials *azcredentials.AzureManagedIdentityCredentials) TokenCredential {
```

Comment:
not related to this PR but I find a bit confusing to have `TokenCredential` and `azcore.TokenCredential` both being different interfaces. As far as I can tell the difference is that this `TokenCredential` manages a cache, so I'd suggest to use `CachedTokenCredential` to make it clear.

Reply:
Agree, the naming is confusing.

Initially I wrote the `ConcurrentTokenCache` as generic cache, to related to Azure SDK.
Another thing is initialization of `azcore.TokenCredential` is expensive, so I wanted a lightweight wrapper with lazy initialization via `Init`.

I will rename it.

Reply:
Renamed to `TokenRetriever`. Wanted to call `TokenAccessor` but `AccessToken` vs `TokenAccessor` sounds weird :)
I decided to not call it `CachedTokenCredential` because it actually doesn't cache the token, it's accessor / getter. Maybe could call `getter` but doesn't metter, it's internal implementation which is not exposed out of the package,